core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java

```diff
@@ -0,0 +1,280 @@
+    private static final ConnectionPoolListener DEFAULT_CONNECTION_POOL_LISTENER =
+            ConnectionPoolListener.noop();
+
+    // Do not accept 1) the options that may break Armeria and 2) the deprecated options.
+    @SuppressWarnings("deprecation")
+    private static final Set<ChannelOption<?>> PROHIBITED_SOCKET_OPTIONS = ImmutableSet.of(
+            ChannelOption.ALLOW_HALF_CLOSURE, ChannelOption.AUTO_READ,
+            ChannelOption.AUTO_CLOSE, ChannelOption.MAX_MESSAGES_PER_READ,
+            ChannelOption.WRITE_BUFFER_HIGH_WATER_MARK, ChannelOption.WRITE_BUFFER_LOW_WATER_MARK,
+            EpollChannelOption.EPOLL_MODE);
+
+    private static final ClientFactoryOptionValue<?>[] DEFAULT_OPTIONS = {
+            ClientFactoryOption.WORKER_GROUP.newValue(DEFAULT_WORKER_GROUP),
+            ClientFactoryOption.SHUTDOWN_WORKER_GROUP_ON_CLOSE.newValue(false),
+            ClientFactoryOption.EVENT_LOOP_SCHEDULER.newValue(DEFAULT_EVENT_LOOP_SCHEDULER),
+            ClientFactoryOption.CHANNEL_OPTIONS.newValue(new Object2ObjectArrayMap<>()),
+            ClientFactoryOption.SSL_CONTEXT_CUSTOMIZER.newValue(DEFAULT_SSL_CONTEXT_CUSTOMIZER),
+            ClientFactoryOption.ADDRESS_RESOLVER_GROUP.newValue(DEFAULT_ADDRESS_RESOLVER_GROUP),
+            ClientFactoryOption.HTTP2_INITIAL_CONNECTION_WINDOW_SIZE.newValue(
+                    Flags.defaultHttp2InitialConnectionWindowSize()),
+            ClientFactoryOption.HTTP2_INITIAL_STREAM_WINDOW_SIZE.newValue(
+                    Flags.defaultHttp2InitialStreamWindowSize()),
+            ClientFactoryOption.HTTP2_MAX_FRAME_SIZE.newValue(Flags.defaultHttp2MaxFrameSize()),
+            ClientFactoryOption.HTTP2_MAX_HEADER_LIST_SIZE.newValue(Flags.defaultHttp2MaxHeaderListSize()),
+            ClientFactoryOption.HTTP1_MAX_INITIAL_LINE_LENGTH.newValue(
+                    Flags.defaultHttp1MaxInitialLineLength()),
+            ClientFactoryOption.HTTP1_MAX_HEADER_SIZE.newValue(Flags.defaultHttp1MaxHeaderSize()),
+            ClientFactoryOption.HTTP1_MAX_CHUNK_SIZE.newValue(Flags.defaultHttp1MaxChunkSize()),
+            ClientFactoryOption.IDLE_TIMEOUT_MILLIS.newValue(Flags.defaultClientIdleTimeoutMillis()),
+            ClientFactoryOption.USE_HTTP2_PREFACE.newValue(Flags.defaultUseHttp2Preface()),
+            ClientFactoryOption.USE_HTTP1_PIPELINING.newValue(Flags.defaultUseHttp1Pipelining()),
+            ClientFactoryOption.CONNECTION_POOL_LISTENER.newValue(DEFAULT_CONNECTION_POOL_LISTENER),
+            ClientFactoryOption.METER_REGISTRY.newValue(Metrics.globalRegistry)
+    };
+
+    public static final ClientFactoryOptions DEFAULT = new ClientFactoryOptions(DEFAULT_OPTIONS);
+
+    /**
+     * TBW.
+     */
+    public static ClientFactoryOptions of(ClientFactoryOptionValue<?>... options) {
+        requireNonNull(options, "options");
+        if (options.length == 0) {
+            return DEFAULT;
+        }
+        return new ClientFactoryOptions(DEFAULT, options);
```

Comment:
Could add `of(ClientFactoryOptions baseOptions, ...)` and simplify into `return of(of(), options)`. Please make this class have the same set of static factory methods with `ClientOptions`.